File: app/src/main/java/me/yeojoy/hancahouse/app/adapter/HouseAdapter.java
package me.yeojoy.hancahouse.app.adapter;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import me.yeojoy.hancahouse.BuildConfig;
import me.yeojoy.hancahouse.MainContract;
import me.yeojoy.hancahouse.MainPresenter;
import me.yeojoy.hancahouse.R;
import me.yeojoy.hancahouse.app.Constants;
import me.yeojoy.hancahouse.app.GlideApp;
import me.yeojoy.hancahouse.model.House;
import me.yeojoy.hancahouse.view.MainView;

public class HouseAdapter extends RecyclerView.Adapter<HouseAdapter.HouseViewHolder> {

    private List<House> mHouses;
    private MainView mMainView;

    public HouseAdapter(MainContract.Presenter presenter) {
        mHouses = new ArrayList<>();
//        mMainView = mainView;
    }

    @NonNull
    @Override
    public HouseViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.row_house, viewGroup, false);
        return new HouseViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull HouseViewHolder holder, int i) {
        House house = mHouses.get(i);

        if (!TextUtils.isEmpty(house.getThumbnailUrl()) &&
                !Constants.NO_IMAGE.equals(house.getThumbnailUrl())) {
            GlideApp.with(holder.mImageViewThumbnail)
                    .load(house.getThumbnailUrl())
                    .fitCenter()
                    // icon from https://www.brandeps.com/icon/H/Home-01-thanks?filetype=zip
                    .placeholder(R.drawable.ic_home_black_24dp)
                    .into(holder.mImageViewThumbnail);
        } else {
            holder.mImageViewThumbnail.setImageResource(R.drawable.ic_home_black_24dp);
        }

        SimpleDateFormat simpleDateFormat
                = new SimpleDateFormat(Constants.APP_DATE_FORMAT, Locale.getDefault());
        holder.mTextViewTitle.setText(house.getTitle());
        holder.mTextViewDate.setText(simpleDateFormat.format(new Date(house.getDate())));

        String author = house.getAuthor() == null ? "" : house.getAuthor();
        if (BuildConfig.DEBUG) {
            author = author + " >>> uid : " + String.valueOf(house.getUid());
        }

        holder.mTextViewAuthor.setText(author);

        holder.itemView.setOnClickListener(view -> mMainView.onItemClick(holder.mImageViewThumbnail, house));
    }

    @Override
    public int getItemCount() {
        return mHouses.size();
    }

    public void setHouses(List<House> houses) {
        mHouses.clear();
        mHouses.addAll(houses);
        notifyDataSetChanged();
    }

    static class HouseViewHolder extends RecyclerView.ViewHolder {

        ImageView mImageViewThumbnail;
        TextView mTextViewTitle, mTextViewDate, mTextViewAuthor;

        HouseViewHolder(@NonNull View itemView) {
            super(itemView);

            mImageViewThumbnail = itemView.findViewById(R.id.image_view_thumbnail);
            mTextViewTitle = itemView.findViewById(R.id.text_view_title);
            mTextViewDate = itemView.findViewById(R.id.text_view_date);
            mTextViewAuthor = itemView.findViewById(R.id.text_view_author);
        }
    }
}

File: app/src/main/java/me/yeojoy/hancahouse/app/adapter/HouseAdapter.kt
package me.yeojoy.hancahouse.app.adapter;

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import me.yeojoy.hancahouse.BuildConfig
import me.yeojoy.hancahouse.MainContract
import me.yeojoy.hancahouse.R
import me.yeojoy.hancahouse.app.Constants
import java.text.SimpleDateFormat
import java.util.*

class HouseAdapter(private val presenter: MainContract.Presenter)
    : RecyclerView.Adapter<HouseAdapter.HouseViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HouseViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.row_house, parent, false)
        return HouseViewHolder(view, presenter::houseItemClicked)
    }

    override fun onBindViewHolder(holder: HouseViewHolder, position: Int) {
        val house = presenter.houseAt(position)
        if (house.thumbnailUrl.isNotEmpty() && !house.thumbnailUrl.equals(Constants.NO_IMAGE)) {
            Glide.with(holder.itemView)
                    .load(house.thumbnailUrl)
                    .fitCenter()
                    .placeholder(R.drawable.ic_home_black_24dp)
                    .into(holder.imageViewThumbnail)
        } else {
            holder.imageViewThumbnail.setImageResource(R.drawable.ic_home_black_24dp)
        }

        val simpleDateFormat = SimpleDateFormat(Constants.APP_DATE_FORMAT, Locale.getDefault())
        holder.textViewTitle.text = house.title
        holder.textViewDate.text = simpleDateFormat.format(Date(house.date))

        val author = if (BuildConfig.DEBUG) {
            "${house.author} >>> uid : ${house.uid.toString()}"
        } else {
            house.author
        }

        holder.textViewAuthor.text = author
    }

    override fun getItemCount(): Int {
        return presenter.numberOfHouses()
    }

    class HouseViewHolder(itemView: View, val itemClickListener: (position: Int) -> Unit)
        : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        init {
            itemView.setOnClickListener(this)
        }

        val imageViewThumbnail: ImageView = itemView.findViewById(R.id.image_view_thumbnail)
        val textViewTitle: TextView = itemView.findViewById(R.id.text_view_title)
        val textViewDate: TextView = itemView.findViewById(R.id.text_view_date)
        val textViewAuthor: TextView = itemView.findViewById(R.id.text_view_author)

        override fun onClick(v: View?) {
            itemClickListener(adapterPosition)
        }
    }
}
